packages/contracts-bedrock/scripts/upgrades/holocene/DeployUpgrade.s.sol

```diff
@@ -0,0 +1,311 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.15;
+
+// Forge
+import { console2 as console } from "forge-std/console2.sol";
+
+// Scripts
+import { Deployer } from "scripts/deploy/Deployer.sol";
+import { DeployUtils } from "scripts/libraries/DeployUtils.sol";
+
+// Utils
+import "src/dispute/lib/Types.sol";
+
+// Interfaces
+import { ISystemConfig } from "src/L1/interfaces/ISystemConfig.sol";
+import { ISuperchainConfig } from "src/L1/interfaces/ISuperchainConfig.sol";
+import { IProxy } from "src/universal/interfaces/IProxy.sol";
+import {
+    IFaultDisputeGame,
+    IBigStepper,
+    IAnchorStateRegistry,
+    IDelayedWETH
+} from "src/dispute/interfaces/IFaultDisputeGame.sol";
+import { IPermissionedDisputeGame } from "src/dispute/interfaces/IPermissionedDisputeGame.sol";
+import { IMIPS, IPreimageOracle } from "src/cannon/interfaces/IMIPS.sol";
+
+/// @title DeployUpgrade
+/// @notice A deployment script for smart contract upgrades surrounding the Holocene hardfork.
+contract DeployUpgrade is Deployer {
+    /// @dev The entrypoint to the deployment script.
+    function deploy(
+        address _proxyAdmin,
+        address _superchainConfig,
+        address _systemConfigImpl,
+        address _mipsImpl,
+        address _delayedWETH,
+        address _preimageOracle,
+        address _anchorStateRegistry
+    )
+        public
+    {
+        // Shim the existing contracts that this upgrade is dependent on.
+        shim({
+            _proxyAdmin: _proxyAdmin,
+            _superchainConfig: _superchainConfig,
+            _systemConfigImpl: _systemConfigImpl,
+            _mipsImpl: _mipsImpl,
+            _delayedWETH: _delayedWETH,
+            _preimageOracle: _preimageOracle,
+            _anchorStateRegistry: _anchorStateRegistry
+        });
+
+        // Deploy new implementations.
+        if (_systemConfigImpl == address(0)) deploySystemConfigImplementation();
+        if (_mipsImpl == address(0)) deployMIPSImplementation();
+        if (_delayedWETH == address(0)) deployDelayedWETH();
+        deployFaultDisputeGameImplementation();
```

Comment:
Strictly speaking, chains that don't use permissionless proofs don't need to deploy the `FaultDisputeGame` contract.  Though I'm not against always deploying it and then choosing whether or not to activate it with a call to `setImplementation`.

Reply:
See https://github.com/ethereum-optimism/optimism/pull/12878#discussion_r1837124140 - I needed a way for the user to specify whether or not they wanted to use the FDG anyways, so was able to make the deployment of the `FaultDisputeGame` and the `DelayedWETH` proxy for it conditional w/o hassle.